### ISSUE TITLE
Use latest Bundler or version from `Gemfile.lock`?

### ISSUE THREAD

**Issue by mokagio** [Automattic/bash-cache-buildkite-plugin#16]
## Background

While working on [adding a pipeline to rebuild the pods and Git cache on a schedule for Pocket Casts iOS](https://github.com/shiftyjelly/pocketcasts-ios/pull/3667/) I run into what looks like a Bundler related issue:

```
bundler: failed to load command: pod (/usr/local/var/buildkite-agent/builds/builder/wordpress-mobile/pocket-casts-ios/vendor/bundle/ruby/2.7.0/bin/pod)

Gem::LoadError: You have already activated bundler 2.2.28, but your Gemfile requires bundler 2.1.4. Since bundler is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports bundler as a default gem.
```

See:

- [The first time I got the error](https://buildkite.com/wordpress-mobile/pocket-casts-ios/builds/7434#2577bfdc-187a-47e0-a415-90dd8415c960/711-712)
- [The build that gets the error again after I reverted the fix as a test](https://buildkite.com/wordpress-mobile/pocket-casts-ios/builds/7444#4864878c-7ae6-4826-8ec6-a7048e3a5593/536-537)

The only way I managed to work around it was by calling `gem install bundler:2.2.32` (the latest version at the time of writing) before calling `install_gems`. Green build [here](https://buildkite.com/wordpress-mobile/pocket-casts-ios/builds/7441).

The issue, as far as I can place it, is that the machine is running on a 2.1.x version of Bundler.

## Suggestion / Question

I wonder if it would be safe, or useful, to parse the `Gemfile.lock` for the `BUNDLE WITH` value and install that version of Bundler before running other commands in those actions that depend of `bundle exec` 🤔 

Alternatively, if we can prove the issue is just a matter of `2.1.x` vs `2.2.x` quirks, we might want to ensure we run on the latest `2.2.x` as part as our VMs boot process (which is of course out of scope for this repo).

**Comment by AliSoftware** [comment]
I wonder if we shouldn't just include `gem install bundler` as part of [the `install_gems` command of this repo](https://github.com/Automattic/bash-cache-buildkite-plugin/blob/trunk/bin/install_gems), to fix it once and for all everywhere automatically?

\cc @jkmassel @mokagio wdyt?

**Comment by mokagio** [comment]
> I wonder if we shouldn't just include gem install bundler as part of [the install_gems command of this repo](https://github.com/Automattic/bash-cache-buildkite-plugin/blob/trunk/bin/install_gems?rgh-link-date=2022-06-23T08%3A21%3A53Z),

That would be an excellent workaround. And, realistically, it wouldn't slow down the pipelines much when compared with the time a build takes.

I have zero knowledge of the process of creating an image for the VMs, but I was hoping we could fix the issue at that "root" level. [See the issues from the buildkite-ci repo linked above](79-gh-Automattic/buildkite-ci):

> I first reported the issue in the bash-cache Buildkite plugin repo, Automattic/bash-cache-buildkite-plugin#16, but after chatting with @spencertransier on Slack, this project seems to be the best place where to implement a fix.

Nothing stops us from addressing it here first as a workaround, though, given it'll be a couple of months at least before we'll build a new VM image. (I'm half tempted to go off on a tangent and create one for Xcode 14.0 beta, but I've already got too many tangents...)

**Comment by AliSoftware** [comment]
> That would be an excellent workaround.

Implemented in https://github.com/Automattic/bash-cache-buildkite-plugin/pull/28

**Comment by jkmassel** [comment]
IINM @rynaardb fixed this on the macOS VM – am I remembering right?

**Comment by mokagio** [comment]
If there's been a fix, maybe it's yet to be deployed? I experienced the issue just a few days ago in Pocket Casts iOS.

Update: Here's a fresh example of the issue with the latest Xcode image, 13.4.1: https://buildkite.com/automattic/wordpress-ui-ios/builds/51#0181fac9-90c6-41fa-8097-112c79250137

**Comment by rynaardb** [comment]
Yep, it seems like it might be the same issue that I've addressed in https://github.com/Automattic/buildkite-ci/pull/77. But you are correct @mokagio, we haven't yet provisioned a new VM with the fix. Up until now, we temporarily "fixed" it at the pipeline level for Tumblr. Have a look over [here](https://github.tumblr.net/TumblrMobile/orangina/blob/492ec11b00fb5220637b7772cc77c54648f125da/.buildkite/commands/tests-components.sh#L3-L8) for example. 

@jkmassel In your [comment](https://github.com/Automattic/buildkite-ci/pull/77#pullrequestreview-998962728) you specifically asked that we work around the issue without having to re-provision a new VM. But it seems like now might be a good time to maybe provision a new VM as it no longer just causing issues for Tumblr but also for other projects. WDYT?

**Comment by mokagio** [comment]
My two cents: We worked around it everywhere I think. In 6-8 weeks, we'll have to provision new VMs for Xcode 14.0. We might as well wait.

@crazytonyli might be looking into making images for Xcode 14.0 betas, so we might have a way to verify this even sooner.

**Comment by mokagio** [comment]
To followup on my [previous comment](https://github.com/Automattic/bash-cache-buildkite-plugin/issues/16#issuecomment-1188216245). It's now been 6-8 weeks, @crazytonyli made new images and also tested removing the `gem install bundler` call [from WordPress iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/19311).

Unfortunately, the issue seems to still be there:

![image](https://user-images.githubusercontent.com/1218433/190311821-9158076f-33cd-41b5-a072-05e20d07ea4a.png)

**Comment by AliSoftware** [comment]
Given this, I think it's worth shipping the workaround for the `install_gems` after all. Especially since that workaround shouldn't really hurt execution time, as installing a `bundler` version that is already installed is a no-op that exits early.

If we don't like using the latest `bundler` version like my fix from #28 does, and prefer to install the version of bundler defined in the `Gemfile.lock`, extracting that value should be as easy as:
```sh
BUNDLER_VERSION=$(sed -n -e '/BUNDLED WITH/{n;s/ *//p;}' Gemfile.lock)
gem install bundler -v "$BUNDLER_VERSION"
```
_(The `sed` command searches for a line matching `BUNDLED WITH`, then gets the `n`ext line after it, `s`ubstitute each space with empty string to trim the result, and `p`rints it.)_

**Comment by mokagio** [comment]
> Given this, I think it's worth shipping the workaround [...]

+1


> If we don't like using the latest bundler version [...]

I like using the latest bundle version 😄
